openvdb/points/PointMove.h

```diff
@@ -0,0 +1,997 @@
+    else if (array.template hasValueType<Mat3<double>>())       ArrayProcessor<Mat3<double>, OpType>::call(op, array);
+    else if (array.template hasValueType<Mat4<float>>())        ArrayProcessor<Mat4<float>, OpType>::call(op, array);
+    else if (array.template hasValueType<Mat4<double>>())       ArrayProcessor<Mat4<double>, OpType>::call(op, array);
+    else if (array.template hasValueType<Quat<float>>())        ArrayProcessor<Quat<float>, OpType>::call(op, array);
+    else if (array.template hasValueType<Quat<double>>())       ArrayProcessor<Quat<double>, OpType>::call(op, array);
+    else                                                        return false;
+    return true;
+}
+
+
+/// Cache read and write attribute handles to amortize construction cost
+struct AttributeHandles
+{
+    using HandleArray = std::vector<AttributeHandle<int>::Ptr>;
+
+    AttributeHandles(const size_t size)
+        : mHandles() { mHandles.reserve(size); }
+
+    AttributeArray& getArray(const Index leafOffset)
+    {
+        auto* handle = reinterpret_cast<AttributeWriteHandle<int>*>(mHandles[leafOffset].get());
+        assert(handle);
+        return handle->array();
+    }
+
+    const AttributeArray& getConstArray(const Index leafOffset) const
+    {
+        const auto* handle = mHandles[leafOffset].get();
+        assert(handle);
+        return handle->array();
+    }
+
+    template <typename ValueT>
+    AttributeHandle<ValueT>& getHandle(const Index leafOffset)
+    {
+        auto* handle = reinterpret_cast<AttributeHandle<ValueT>*>(mHandles[leafOffset].get());
+        assert(handle);
+        return *handle;
+    }
+
+    template <typename ValueT>
+    AttributeWriteHandle<ValueT>& getWriteHandle(const Index leafOffset)
+    {
+        auto* handle = reinterpret_cast<AttributeWriteHandle<ValueT>*>(mHandles[leafOffset].get());
+        assert(handle);
+        return *handle;
+    }
+
+    /// Create a handle and reinterpret cast as an int handle to store
+    struct CacheHandleOp
+    {
+        CacheHandleOp(HandleArray& handles)
+            : mHandles(handles) { }
+
+        template<typename ValueT>
+        void operator()(const AttributeArray& array) const
+        {
+            auto* handleAsInt = reinterpret_cast<AttributeHandle<int>*>(
+                new AttributeHandle<ValueT>(array));
+            mHandles.emplace_back(handleAsInt);
+        }
+
+    private:
+        HandleArray& mHandles;
+    }; // struct CacheHandleOp
+
+    template <typename LeafRangeT>
+    void cache(const LeafRangeT& range, const Index attributeIndex)
+    {
+        using namespace openvdb::math;
+
+        mHandles.clear();
+        CacheHandleOp op(mHandles);
+
+        for (auto leaf = range.begin(); leaf; ++leaf) {
+            auto& array = leaf->attributeArray(attributeIndex);
+            processTypedArray(array, op);
+        }
+    }
+
+private:
+    HandleArray mHandles;
```

Comment:
Absolutely, however it will require an ABI change to be able do this properly. See the comment discussing this in detail here - https://github.com/dreamworksanimation/openvdb/blob/master/openvdb/points/PointMove.h#L197

I intend to address this deficiency in the AttributeArray API for ABI=6 at which stage lots of this framework can be deprecated.